### PR TITLE
Mailer library

### PR DESCRIPTION
**Pull request by armanist**

### CODE REVIEW

File: src/Libraries/Storage/Adapters/Local/LocalFileSystemAdapter.php [hunk 1]
```diff
@@ -203,12 +203,17 @@ public function isWritable(string $filename): bool
      * @param string $filename
      * @param int $offset
      * @param int|null $length
-     * @param int $flags
      * @return array
      */
-    public function getLines(string $filename, int $offset, ?int $length, int $flags = 0): array
+    public function getLines(string $filename, int $offset = 0, ?int $length = null): array
```
Comment (Vahram1995) [inline src/Libraries/Storage/Adapters/Local/LocalFileSystemAdapter.php:208]: method can return `false`

**Review by Vahram1995 — CHANGES_REQUESTED** [review]

**Review by andrey-smaelov — APPROVED** [review]

**Review by hatik92 — APPROVED** [review]

File: src/Libraries/Mailer/Adapters/MandrillAdapter.php [hunk 2]
```diff
@@ -16,29 +16,128 @@
...
+    private function sendEmail(): bool
     {
         try {
-            $data["key"] = config()->get('mailer.mandrill.api_key');
-
-            $httpClient = new HttpClient();
-            $httpClient->createMultiRequest()
-                ->createRequest("https://mandrillapp.com/api/1.0/messages/send.json")
-                ->setMethod("POST")
-                ->setData($data)
```
Comment (hatik92) [inline src/Libraries/Mailer/Adapters/MandrillAdapter.php:36]: for mandrill api_key should be in data
$this->data["key"] = $this->apiKey;

File: src/Libraries/Mailer/Adapters/SendgridAdapter.php [hunk 3]
```diff
@@ -16,30 +16,131 @@
...
+            if ($this->templatePath) {
+                $body = $this->createFromTemplate();
+            } else {
+                $body = is_array($this->message) ? implode($this->message) : $this->message;
+            }
+
+            $this->data['content'] = [
+                'type' => 'text/html',
+                'value' => trim(str_replace("\n", "", $body))
+            ];
```
Comment (hatik92) [inline src/Libraries/Mailer/Adapters/SendgridAdapter.php:112]: data['content '] should be in array

$this->data['content'] = [
                [
                    'type' => 'text/html',
                    'value' => trim(str_replace("\n", "", $body))
                ]
            ];

File: src/Libraries/Mailer/Adapters/SendgridAdapter.php [hunk 4]
```diff
@@ -16,30 +16,131 @@
...
+        return self::$instance;
+    }
+
+    /**
+     * Prepares the data
+     */
+    private function prepare()
+    {
+        $this->data['from'] = [
+            'email' => $this->from['email']
```
Comment (hatik92) [inline src/Libraries/Mailer/Adapters/SendgridAdapter.php:91]: add to data from
'name' => $this->from['name']

File: src/Libraries/Mailer/Adapters/SmtpAdapter.php [hunk 5]
```diff
@@ -15,17 +15,299 @@
...
+     */
+    private function setupSmtp(array $params)
+    {
+        $this->mailer->isSMTP();
+        $this->mailer->SMTPAuth = true;
+        $this->mailer->Host = $params['host'];
+        $this->mailer->SMTPSecure = $params['secure'];
+        $this->mailer->Port = $params['port'];
+        $this->mailer->Username = $params['username'];
+        $this->mailer->Password = $params['password'];
```
Comment (hatik92) [inline src/Libraries/Mailer/Adapters/SmtpAdapter.php:239]: params keys add mail_

File: src/Libraries/Mailer/Adapters/SmtpAdapter.php [hunk 6]
```diff
@@ -15,17 +15,299 @@
...
+        }
+
+        if ($this->message) {
+            if ($this->templatePath) {
+                $body = $this->createFromTemplate();
+            } else {
+                $body = is_array($this->message) ? implode($this->message) : $this->message;
+            }
+
+            $this->mailer->Body = $body;
```
Comment (hatik92) [inline src/Libraries/Mailer/Adapters/SmtpAdapter.php:261]: replase
$this->mailer->Body = trim(str_replace("\n", "", $body));

File: tests/unit/_root/shared/config/mailer.php [hunk 7]
```diff
@@ -5,7 +5,13 @@
 
     'mail_trap' => true,
 
-    'smtp' => [],
+    'smtp' => [
+        'host' => '127.0.0.1',
+        'secure' => 'ssl',
+        'port' => '80',
+        'username' => 'test',
+        'password' => 'test'
```
Comment (hatik92) [inline tests/unit/_root/shared/config/mailer.php:13]: keys add mail_

**Review by hatik92 — APPROVED** [review]

**Review by hatik92 — APPROVED** [review]

**Review by andrey-smaelov — APPROVED** [review]

**Pull request merged**